### PR TITLE
Aurora/Hestia-Inspired Cooking System

### PR DESCRIPTION
**Pull request by MarinaGryphon**
A port of Hestia's cooking system with some backported features from Aurora. No new foods or recipes have been added, but the current recipes have been redone.

TODO before completion:
- [ ] Add new items to kitchen vendor
- [ ] Port ladles and chopsticks
- [ ] Test beating people up with skillets
- [ ] Large test with friends
- [ ] Get around to implementing world/inventory/worn icons for saucepans

## Changelog
:cl: Jade2562
tweak: Completely overhauled recipe system and recipes
soundadd: Random sound loops, used by the fryer and microwave
rscadd: New microwave UI with an animated progress bar
tweak: Completely overhauled the oven and fryer
rscadd: Added the stove, which also heats reagents and has dynamic overlays displaying its contents! Nifty.
rscadd: Recipes that can be done in more than one appliance *without* being duplicated
rscadd: Cooking containers, used in certain appliances to cook things
rscadd: Appliances that can take more than one type of cooking container
rscadd: The ability to beat people with frying pans!
rscadd: The ability to wear saucepans
rscdel: Removed the grill, to be re-implemented at a later date.
imageadd: Added some nice new sprites for appliances and the microwave
imagedel: deleted some icons and images
maptweak: Replaced grills on the map with stoves.
/:cl:

### CODE REVIEW

File: code/modules/cooking/machinery/cooking_machines/_appliance.dm [hunk 1]
```diff
@@ -0,0 +1,653 @@
...
+	var/combine_first = FALSE//If 1, this appliance will do combination cooking before checking recipes
+
+/obj/machinery/appliance/Initialize()
+	. = ..()
+	if(length(output_options))
+		verbs += /obj/machinery/appliance/proc/choose_output
+	if(powered())
+		stat &= ~NOPOWER
+	else
+		stat |= NOPOWER
```
Comment (afterthought2) [inline code/modules/cooking/machinery/cooking_machines/_appliance.dm]: Do not do this kind of stuff. Power is handled for you. Do not modify the stat var directly, ever.

File: code/modules/cooking/machinery/cooking_machines/_appliance.dm [hunk 2]
```diff
@@ -0,0 +1,653 @@
...
+	for (var/a in cooking_objs)
+		var/datum/cooking_item/CI = a
+		qdel(CI.container)//Food is fragile, it probably doesnt survive the destruction of the machine
+		cooking_objs -= CI
+		qdel(CI)
+	return ..()
+
+/obj/machinery/appliance/examine(var/mob/user)
+	..()
+	if(Adjacent(usr))
```
Comment (afterthought2) [inline code/modules/cooking/machinery/cooking_machines/_appliance.dm]: `usr`, and also incorrect format.

File: code/modules/cooking/machinery/cooking_machines/_appliance.dm [hunk 3]
```diff
@@ -0,0 +1,653 @@
...
+		return
+
+	if (user.stat || user.restrained() || user.incapacitated())
+		return
+
+	if (!Adjacent(user) && !issilicon(user))
+		to_chat(user, "You can't reach [src] from here.")
+		return
+
+	use_power = (use_power == POWER_USE_OFF) ? POWER_USE_IDLE : POWER_USE_OFF // If on, use active power, else use no power
```
Comment (afterthought2) [inline code/modules/cooking/machinery/cooking_machines/_appliance.dm]: Do not set this var directly.

File: code/modules/cooking/machinery/cooking_machines/_appliance.dm [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+	return TRUE
+
+/obj/machinery/appliance/attackby(var/obj/item/I, var/mob/user)
+	if(!cook_type || (stat & (BROKEN)))
+		to_chat(user, SPAN_WARNING("[src] is not working."))
+		return
+
+	var/result = can_insert(I, user)
+	if(result == CANNOT_INSERT)
+		return component_attackby(I, user)
```
Comment (afterthought2) [inline code/modules/cooking/machinery/cooking_machines/_appliance.dm]: It's likely better to return `..()` unless there's a reason not to.

File: code/modules/cooking/machinery/cooking_machines/_appliance.dm [hunk 5]
```diff
@@ -0,0 +1,653 @@
...
+/obj/machinery/appliance/CtrlClick(var/mob/user)
+	if(!anchored)
+		return ..()
+	attempt_toggle_power(user)
+
+/obj/machinery/appliance/attack_hand(var/mob/user)
+	if (!length(cooking_objs))
+		return
+	if (removal_menu(user))
+		return
```
Comment (afterthought2) [inline code/modules/cooking/machinery/cooking_machines/_appliance.dm:507]: Put this in `physical_attack_hand`.

File: code/modules/cooking/machinery/cooking_machines/_cooker.dm [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+/obj/machinery/appliance/cooker/Process()
+	if ((temperature >= set_temp) && (stat || use_power == 1))
+		QUEUE_TEMPERATURE_ATOMS(src) // cool every tick if we're not turned on or heating
+	if(!stat)
+		heat_up()
+	. = ..()
+
+/obj/machinery/appliance/cooker/power_change()
+	. = ..()
+	queue_icon_update()
```
Comment (afterthought2) [inline code/modules/cooking/machinery/cooking_machines/_cooker.dm]: It already does this if needed.

File: code/modules/cooking/machinery/cooking_machines/_mixer.dm [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+/obj/machinery/appliance/mixer/on_update_icon()
+	if (!stat)
+		icon_state = on_icon
+	else
+		icon_state = off_icon
+
+
+/obj/machinery/appliance/mixer/Process()
+	if (!stat)
```
Comment (afterthought2) [inline code/modules/cooking/machinery/cooking_machines/_mixer.dm]: Check the actual stat flags you care about.

File: code/modules/cooking/machinery/cooking_machines/_mixer.dm [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+/obj/machinery/appliance/mixer/finish_cooking(var/datum/cooking_item/CI)
+	..()
+	playsound(src, 'sound/machines/click.ogg', 40, 1)
+	update_use_power(POWER_USE_OFF)
+	CI.reset()
+	update_icon()
+
+/obj/machinery/appliance/mixer/on_update_icon()
+	if (!stat)
```
Comment (afterthought2) [inline code/modules/cooking/machinery/cooking_machines/_mixer.dm]: Same here, suppressing further comments about this.

File: code/modules/cooking/machinery/cooking_machines/_mixer.dm [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+	set category = "Object"
+
+	if (!isliving(usr))
+		return
+
+	if (usr.incapacitated(usr))
+		return
+
+	if (!usr.check_dexterity(DEXTERITY_SIMPLE_MACHINES))
+		return
```
Comment (afterthought2) [inline code/modules/cooking/machinery/cooking_machines/_mixer.dm:41]: `CanPhysicallyInteract`.

File: code/modules/cooking/machinery/cooking_machines/oven.dm [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+		temperature_coefficient = 10
+		//When the oven door is opened, oven loses heat faster
+
+	playsound(src, 'sound/machines/hatch_open.ogg', 20, 1)
+	update_icon()
+
+/obj/machinery/appliance/cooker/oven/proc/manip(var/obj/item/I)
+	// check if someone's trying to manipulate the machine
+
+	return isCrowbar(I) || isScrewdriver(I) || istype(I, /obj/item/storage/part_replacer) || istype(I, /obj/item/stock_parts)
```
Comment (afterthought2) [inline code/modules/cooking/machinery/cooking_machines/oven.dm]: The correct way to handle this is to call `component_interact` first, and if nothing happened, proceed to insertion.

File: code/modules/cooking/machinery/cooking_machines/stove.dm [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+	starts_with = list(
+		/obj/item/chems/cooking_container/skillet,
+		/obj/item/chems/cooking_container/pot,
+		/obj/item/chems/cooking_container/saucepan
+	)
+
+/obj/machinery/appliance/cooker/stove/on_update_icon()
+	. = ..()
+	cut_overlays()
```
Comment (afterthought2) [inline code/modules/cooking/machinery/cooking_machines/stove.dm]: This is not compatible with calling `..()`.

File: code/modules/cooking/machinery/microwave.dm [hunk 12]
```diff
@@ -0,0 +1,477 @@
...
+	soundloop = new(list(src), FALSE)
+
+	RefreshParts()
+
+/*******************
+*   Item Adding
+********************/
+
+/obj/machinery/microwave/proc/add_item(var/obj/item/W as obj, var/mob/user as mob)
+	user.drop_from_inventory(W, src)
```
Comment (afterthought2) [inline code/modules/cooking/machinery/microwave.dm]: It is almost never correct to call this proc.

File: code/modules/cooking/machinery/microwave.dm [hunk 13]
```diff
@@ -0,0 +1,477 @@
...
+					SPAN_NOTICE("\The [user] fixes the microwave."), \
+					SPAN_NOTICE("You have fixed the microwave.") \
+				)
+				icon_state = "mw"
+				broken = 0 // Fix it!
+				dirty = 0 // just to be sure
+				atom_flags |= ATOM_FLAG_OPEN_CONTAINER | ATOM_FLAG_NO_REACT
+		else
+			to_chat(user, SPAN_WARNING("It's broken!"))
+			return 1
```
Comment (afterthought2) [inline code/modules/cooking/machinery/microwave.dm]: This stuff should be handled at `/machinery` level or not at all.

File: code/modules/cooking/machinery/microwave.dm [hunk 14]
```diff
@@ -0,0 +1,477 @@
...
+	else if((. = component_attackby(O, user)))
+		eject(user, FALSE)
```
Comment (afterthought2) [inline code/modules/cooking/machinery/microwave.dm]: Why? You don't know what actually happened in this proc; you aren't allowed to do random other stuff afterwards. It's possible the machine no longer exists, for example.

File: code/modules/cooking/machinery/microwave.dm [hunk 15]
```diff
@@ -0,0 +1,477 @@
...
+			SPAN_NOTICE("You attempt to [src.anchored ? "unsecure" : "secure"] the microwave.")
+			)
+		if (do_after(user, 2 SECONDS))
+			user.visible_message( \
+			SPAN_NOTICE("\The [user] [src.anchored ? "unsecures" : "secures"] the microwave."), \
+			SPAN_NOTICE("You [src.anchored ? "unsecure" : "secure"] the microwave.")
+			)
+			src.anchored = !src.anchored
+		else
+			to_chat(user, SPAN_NOTICE("You decide not to do that."))
```
Comment (afterthought2) [inline code/modules/cooking/machinery/microwave.dm]: This should be handled at `/obj` level, either via a parent call or via a direct call to the anchorable-flag triggered verb.

File: code/modules/cooking/machinery/microwave.dm [hunk 16]
```diff
@@ -0,0 +1,477 @@
...
+					SPAN_NOTICE("\The [user] has added one of [O] to \the [src]."), \
+					SPAN_NOTICE("You add one of [O] to \the [src]."))
+				SSnano.update_uis(src)
+				return
+			else
+				add_item(O, user)
+		else
+			add_item(O, user)
+			return
+	SSnano.update_uis(src)
```
Comment (afterthought2) [inline code/modules/cooking/machinery/microwave.dm]: Why? Nothing happened.

File: code/modules/cooking/machinery/microwave.dm [hunk 17]
```diff
@@ -0,0 +1,477 @@
...
+	..()
```
Comment (afterthought2) [inline code/modules/cooking/machinery/microwave.dm]: I don't believe the rest of the proc is compatible with a parent call here, but am not sure.

File: code/modules/cooking/machinery/microwave.dm [hunk 18]
```diff
@@ -0,0 +1,477 @@
...
+
+/obj/machinery/microwave/AltClick()
+	if(!operating)
+		cook()
+
+/obj/machinery/microwave/attack_ai(mob/user as mob)
+	if(istype(user, /mob/living/silicon/robot) && Adjacent(user))
+		attack_hand(user)
```
Comment (afterthought2) [inline code/modules/cooking/machinery/microwave.dm]: Please remove; AI overrides are handled at /machinery level, and this one is nonsensical anyway.

File: code/modules/cooking/machinery/microwave.dm [hunk 19]
```diff
@@ -0,0 +1,477 @@
...
+
+/obj/machinery/microwave/attack_hand(mob/user as mob)
+	user.set_machine(src)
```
Comment (afterthought2) [inline code/modules/cooking/machinery/microwave.dm]: Do not call this.

File: code/modules/cooking/machinery/microwave.dm [hunk 20]
```diff
@@ -0,0 +1,477 @@
...
+	if(broken > 0)
+		to_chat(user, SPAN_WARNING("\The [name] is broken! You'll need to fix it before using it."))
+	else if(dirty >= 100)
+		to_chat(user, SPAN_WARNING("\The [name] is dirty! You'll need to clean it before using it."))
+	else
+		ui_interact(user)
```
Comment (afterthought2) [inline code/modules/cooking/machinery/microwave.dm]: Place this in `interface_interact`.

File: code/modules/cooking/machinery/microwave.dm [hunk 21]
```diff
@@ -0,0 +1,477 @@
...
+			// Something's in the microwave that shouldn't be! Time to break!
+			cook_break = TRUE
+	else
+		failed = FALSE
+		cook_time = update_cook_time(round(recipe.time * 2))
+
+	start()
+
+/obj/machinery/microwave/proc/update_cook_time(var/ct = 200)
+	RefreshParts()
```
Comment (afterthought2) [inline code/modules/cooking/machinery/microwave.dm]: Why? The parts haven't changed. You should not be calling this proc manually.

File: code/modules/cooking/machinery/microwave.dm [hunk 22]
```diff
@@ -0,0 +1,477 @@
...
+	return ffuu
+
+/obj/machinery/microwave/OnTopic(var/mob/user, var/href_list, var/datum/topic_state/state)
+	if(dirty >= 100)
+		to_chat(usr, SPAN_WARNING("\The [name] is dirty! You'll need to clean it before using it."))
+		return TOPIC_NOACTION
+
+	if(broken > 0)
+		to_chat(usr, SPAN_WARNING("\The [name] is broken! You'll need to fix it before using it."))
+		return TOPIC_NOACTION
```
Comment (afterthought2) [inline code/modules/cooking/machinery/microwave.dm:383]: These go in `CanUseTopic`, and generally should not return errors (place errors in `/examine`, or if really needed pass hrefs to CanUseTopic).

File: code/modules/cooking/machinery/microwave.dm [hunk 23]
```diff
@@ -0,0 +1,477 @@
...
+	..()
+	var/bin_rating = 0 // 2
+	var/cap_rating = 0 // 3
+	var/las_rating = 0 // 1
+
+	bin_rating = total_component_rating_of_type(/obj/item/stock_parts/matter_bin)
+	cap_rating = total_component_rating_of_type(/obj/item/stock_parts/capacitor)
+	las_rating = total_component_rating_of_type(/obj/item/stock_parts/micro_laser)
+
+	active_power_usage = initial(active_power_usage) - (cap_rating * 25)
```
Comment (afterthought2) [inline code/modules/cooking/machinery/microwave.dm:456]: This var should not be modified directly.

File: code/__defines/misc.dm [hunk 24]
```diff
@@ -266,3 +266,22 @@
 
 #define GET_DECL(D) (ispath(D, /decl) ? (decls_repository.fetched_decls[D] || decls_repository.get_decl(D)) : null)
 #define Z_ALL_TURFS(Z) block(locate(1, 1, Z), locate(world.maxx, world.maxy, Z))
+
+// Cooking appliances.
+#define MIX			BITFLAG(0)
```
Comment (MistakeNot4892) [inline code/__defines/misc.dm]: Please indent with spaces so it doesn't get all wonky in GitHub diff.

File: code/__defines/misc.dm [hunk 25]
```diff
@@ -266,3 +266,22 @@
...
+#define FRYER		BITFLAG(1)
+#define OVEN		BITFLAG(2)
+#define SKILLET		BITFLAG(3)
+#define SAUCEPAN	BITFLAG(4)
+#define POT			BITFLAG(5)
+#define GRILL		BITFLAG(6)
+#define MICROWAVE	BITFLAG(7)
+// Cooking misc.
+// can_insert return values
+#define CANNOT_INSERT		0
```
Comment (MistakeNot4892) [inline code/__defines/misc.dm]: Are these bitflags or enums?

File: code/datums/looping_sounds/_looping_sound.dm [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+
+	end_sound		(soundfile)				The sound played after the main loop has concluded
+
+	chance			(num)					Chance per loop to play a mid_sound
+	volume			(num)					Sound output volume
+	max_loops		(num)					The max amount of loops to run for.
+	direct			(bool)					If true plays directly to provided atoms instead of from them
+*/
+
+/datum/looping_sound
```
Comment (MistakeNot4892) [inline code/datums/looping_sounds/_looping_sound.dm]: As discussed, I think `composite_sound` or `dynamic_sound` would be a less confusing name. I also think it would be good for this to be split into a separate PR for review as it's pretty distinct from the body of the cooking system and could be useful generally even if the cooking system ends up being a poor fit for whatever reason.

File: code/datums/looping_sounds/_looping_sound.dm [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+/datum/looping_sound/proc/play(soundfile)
+	var/list/atoms_cache = output_atoms
+	var/sound/S = sound(soundfile)
+	for(var/i in 1 to atoms_cache.len)
+		var/atom/thing = atoms_cache[i]
+		playsound(thing, S, volume)
+
+/datum/looping_sound/proc/get_sound(starttime, _mid_sounds)
+	. = _mid_sounds || mid_sounds
+	while(!isfile(.) && !isnull(.))
```
Comment (MistakeNot4892) [inline code/datums/looping_sounds/_looping_sound.dm:82]: I don't think isfile works on RSC entries, does it?

File: code/game/machinery/kitchen/smartfridge.dm [hunk 28]
```diff
@@ -156,17 +156,12 @@
 				var/obj/item/chems/food/snacks/S = thing
 				if(S.dry || !I.get_specific_product(get_turf(src), S))
 					continue
-				if(S.dried_type == S.type)
-					S.dry = 1
-					S.SetName("dried [S.name]")
-					S.color = "#a38463"
-					stock_item(S)
+				var/result = S.on_dry(get_turf(src))
```
Comment (MistakeNot4892) [inline code/game/machinery/kitchen/smartfridge.dm:159]: The drying refactor stuff could also be split into a separate PR, anything that improves the mess of a system we have currently is good.

File: code/modules/cooking/machinery/cooking_machines/_cooker.dm [hunk 29]
```diff
@@ -0,0 +1,205 @@
...
+		set_temp = text2num(desired_temp) + T0C
+		to_chat(user, SPAN_NOTICE("You set [src] to [round(set_temp-T0C)]C."))
+		update_use_power(POWER_USE_IDLE) // will be updated later
+	if(old_power != use_power)
+		activation_message(user)
+	playsound(src, 'sound/machines/click.ogg', 40, 1)
+	update_icon()
+
+/obj/machinery/appliance/cooker/proc/activation_message(var/mob/user)
+	user.visible_message("<b>[user]</b> turns [use_power ? "on" : "off"] [src].", "You turn [use_power ? "on" : "off"] [src].")
```
Comment (MistakeNot4892) [inline code/modules/cooking/machinery/cooking_machines/_cooker.dm]: `\The [user]`

File: code/modules/cooking/machinery/cooking_machines/_cooker.dm [hunk 30]
```diff
@@ -0,0 +1,205 @@
...
+
+/obj/machinery/appliance/cooker/on_update_icon()
+	overlays.Cut()
```
Comment (MistakeNot4892) [inline code/modules/cooking/machinery/cooking_machines/_cooker.dm]: Probably worth changing this to use SSoverlays while you're adding it.

File: code/modules/reagents/reagent_containers/food/snacks.dm [hunk 31]
```diff
@@ -179,6 +196,108 @@
...
+
+	return TRUE
+
+//Called by cooking machines. This is mainly intended to set properties on the food that differ between raw/cooked
+/obj/item/chems/food/snacks/proc/cook()
+	if (batter_coating)
+		cut_overlays()
+		var/decl/material/liquid/nutriment/batter/our_coating = decls_repository.get_decl(batter_coating)
+		var/icon/I = icon(icon, icon_state, dir)
+		color = "#FFFFFF" //Some fruits use the color var
```
Comment (MistakeNot4892) [inline code/modules/reagents/reagent_containers/food/snacks.dm]: This will fail CI, use `COLOR_WHITE` or `#ffffff`.

File: code/modules/cooking/machinery/cooking_machines/_cooker.dm (hunk 29 above)
Comment (MarinaGryphon) [reply code/modules/cooking/machinery/cooking_machines/_cooker.dm]: Not necessary (it's auto-inserted), but I'll include it to make it explicit.

File: code/datums/looping_sounds/_looping_sound.dm (hunk 27 above)
Comment (MarinaGryphon) [reply code/datums/looping_sounds/_looping_sound.dm:82]: `This returns a true value when given a file. Both objects returned by file() and files stored in the resource cache qualify.`

File: code/__defines/misc.dm (hunk 25 above)
Comment (MarinaGryphon) [reply code/__defines/misc.dm]: Enums.

File: code/modules/cooking/machinery/microwave.dm (hunk 13 above)
Comment (MarinaGryphon) [reply code/modules/cooking/machinery/microwave.dm]: This is current Nebula code; I'm not really up for a refactor of this right now.

**Comment by MistakeNot4892** [comment]
Closing as discussed on Discord.
